### PR TITLE
JAMES-2150 Added the ability to upload Sieve scripts via the CLI

### PR DESCRIPTION
**Pull request by matzepan**
Implemented the steps described in [JAMES-2150](https://issues.apache.org/jira/browse/JAMES-2150)

### CODE REVIEW

File: server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java [hunk 1]
```diff
@@ -331,6 +331,9 @@ private void executeCommand(String[] arguments, CmdType cmdType, PrintStream pri
         case REMOVESIEVEUSERQUOTA:
             sieveProbe.removeSieveQuota(arguments[1]);
             break;
+            case ADDACTIVESIEVESCRIPT:
```
Comment (rouazana) [inline server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java]: this case should be indented as the previous one

**Comment by chibenwa** [comment]
I am relaunching the CI build on your pull request

test this please

**Comment by matzepan** [comment]
I added the ability to load from a file - but I have to say that I am quite confused what the difference between SieveRepositoryManagement and SieveProbeImpl should be. Looks like quite the duplicate functionality.

File: server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java [hunk 2]
```diff
@@ -125,18 +136,6 @@ public static void executeAndOutputToStream(String[] args, PrintStream printStre
...
-    private final MailboxProbe mailboxProbe;
-    private final QuotaProbe quotaProbe;
-    private final SieveProbe sieveProbe;
-    
-    public ServerCmd(DataProbe probe, MailboxProbe mailboxProbe, QuotaProbe quotaProbe, SieveProbe sieveProbe) {
-        this.probe = probe;
-        this.mailboxProbe = mailboxProbe;
-        this.quotaProbe = quotaProbe;
-        this.sieveProbe = sieveProbe;
-    }
```
Comment (chibenwa) [inline server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java]: That is unfortunate: we tend to put static methods first, then fields, then the constructor, then the methods.

You also moved by mistake in a similar way `print` and `printUsage` below.

Could you restore the original ordering of these methods?

File: server/container/cli/src/main/java/org/apache/james/cli/probe/impl/JmxSieveProbe.java [hunk 3]
```diff
@@ -76,5 +76,10 @@ public void removeSieveQuota(String user) throws Exception {
     public void addActiveSieveScript(String user, String name, String script) throws Exception {
```
Comment (chibenwa) [inline server/container/cli/src/main/java/org/apache/james/cli/probe/impl/JmxSieveProbe.java]: This method is no more needed, right?

File: server/container/guice/protocols/managedsieve/src/main/java/org/apache/james/modules/protocols/SieveProbeImpl.java [hunk 4]
```diff
@@ -68,9 +72,16 @@ public void removeSieveQuota(String user) throws Exception {
...
+    public void addActiveSieveScript(String userName, String name, String script) throws Exception {
+        User user = User.fromUsername(userName);
         sieveRepository.putScript(user, new ScriptName(name), new ScriptContent(script));
         sieveRepository.setActive(user, new ScriptName(name));
     }
+
+    @Override
+    public void addActiveSieveScriptFromFile(String userName, String name, String path) throws Exception {
+        try (InputStream scriptFileAsStream = new FileInputStream(path)) {
+            addActiveSieveScript(userName, name, IOUtils.toString(scriptFileAsStream));
```
Comment (chibenwa) [inline server/container/guice/protocols/managedsieve/src/main/java/org/apache/james/modules/protocols/SieveProbeImpl.java]: `IOUtils.toString(InputStream)` is deprecated.

We need to specify the encoding.

We generally assume that `StandardCharset.UTF-8` does the job.

File: server/data/data-api/src/main/java/org/apache/james/probe/SieveProbe.java [hunk 5]
```diff
@@ -33,6 +33,7 @@
 
     void removeSieveQuota(String user) throws Exception;
 
-    void addActiveSieveScript(String user, String name, String script) throws Exception;
+    void addActiveSieveScript(String userName, String name, String script) throws Exception;
```
Comment (chibenwa) [inline server/data/data-api/src/main/java/org/apache/james/probe/SieveProbe.java]: Is it still used?

File: server/data/data-api/src/main/java/org/apache/james/sieverepository/api/SieveRepositoryManagementMBean.java [hunk 6]
```diff
@@ -38,4 +38,6 @@
 
     void addActiveSieveScript(String userName, String scriptName, String script) throws SieveRepositoryException;
```
Comment (chibenwa) [inline server/data/data-api/src/main/java/org/apache/james/sieverepository/api/SieveRepositoryManagementMBean.java]: Is it still used?

File: server/data/data-library/src/main/java/org/apache/james/sieverepository/lib/SieveRepositoryManagement.java [hunk 7]
```diff
@@ -81,4 +89,13 @@ public void addActiveSieveScript(String userName, String scriptName, String scri
         sieveRepository.putScript(user, new ScriptName(scriptName), new ScriptContent(script));
         sieveRepository.setActive(user, new ScriptName(scriptName));
     }
+
+    @Override
+    public void addActiveSieveScriptFromFile(String userName, String scriptName, String scriptPath) throws SieveRepositoryException {
+        try (InputStream scriptFileAsStream = new FileInputStream(scriptPath)) {
+            addActiveSieveScript(userName, scriptName, IOUtils.toString(scriptFileAsStream));
```
Comment (chibenwa) [inline server/data/data-library/src/main/java/org/apache/james/sieverepository/lib/SieveRepositoryManagement.java]: Idem, you need to explicitly specify the encoding...

File: server/data/data-library/src/test/java/org/apache/james/sieverepository/lib/SieveRepositoryManagementTest.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
+package org.apache.james.sieverepository.lib;
```
Comment (chibenwa) [inline server/data/data-library/src/test/java/org/apache/james/sieverepository/lib/SieveRepositoryManagementTest.java]: You are missing the license... :-)

File: server/data/data-library/src/test/java/org/apache/james/sieverepository/lib/SieveRepositoryManagementTest.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Test
+    public void importSieveScriptFileToRepositoryShouldStoreContentAndActivateScript() throws Exception {
+        String userName = "user@domain";
+        String script = "user_script";
+        URL sieveResource = ClassLoader.getSystemResource("sieve/my_sieve");
+
+        User user = User.fromUsername(userName);
+        ScriptName scriptName = new ScriptName(script);
+        String sieveContent = IOUtils.toString(sieveResource);
```
Comment (chibenwa) [inline server/data/data-library/src/test/java/org/apache/james/sieverepository/lib/SieveRepositoryManagementTest.java]: You miss the encoding...

File: server/data/data-library/src/test/java/org/apache/james/sieverepository/lib/SieveRepositoryManagementTest.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Test
+    public void importSieveScriptFileToRepositoryShouldNotImportFileWithWrongPathToRepistory() throws Exception {
+        String userName = "user@domain";
+        String script = "user_script";
+        URL sieveResource = ClassLoader.getSystemResource("sieve/my_sieve");
+
+        User user = User.fromUsername(userName);
+        ScriptName scriptName = new ScriptName(script);
+        String sieveContent = IOUtils.toString(sieveResource);
```
Comment (chibenwa) [inline server/data/data-library/src/test/java/org/apache/james/sieverepository/lib/SieveRepositoryManagementTest.java]: Here too ...

**Comment by chibenwa** [comment]
```
I added the ability to load from a file - but I have to say that I am quite confused what the difference between SieveRepositoryManagement and SieveProbeImpl should be. Looks like quite the duplicate functionality.
```

Very true.

Note that probes and Management APIs needs to exist: they serve different functionalities:
 - Management and MBean is for some JMX semantic. We do not want it to leak further in the project
 - Probes are a way to interact with a James server. We do rely on them in some tests.

However many of them have duplicate capabilities. Maybe we could have the management stuff calling the probe directly?

File: server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java (hunk 2 above)
Comment (matzepan) [reply server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java]: I did a Rearrange Code in Intellij which i thought used the checkstyle to rearrange the code accordingly. Reverting it :)

**Comment by trantienduchn** [comment]
no build result, test this please

**Comment by chibenwa** [comment]
We recently did some modifications to the merging tool that runs before the pull request build.

In order to benefit from it, and start the tests, you need to rebase on  `linagora/master` then force push your changes here.

Can I let you do it?

Many thanks!

**Comment by matzepan** [comment]
Rebase is complete.

**Comment by chibenwa** [comment]
Hi,

This pull request has just been merged.

Thank you again for your dedication toward the James project. If you need help, want to [contribute](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20newbie%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC) [documentation](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20documentation%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC), new [features](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20feature%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC), tests or [bugfix](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20easyfix%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC), please don't hesitate to reach us.

Thanks again,

Cheers,

Benoit

**Pull request closed without merging**